README.md
### pip-date - Date your pip packages!

[![pypi supported versions][1]][2]
[![PyPI version][3]][4]
[![Maintenance][5]][6]
[![GitHub last commit][7]][8]
[![Average time to resolve an issue][9]][10]

[1]: https://img.shields.io/pypi/pyversions/pip-date.svg
[2]: https://pypi.python.org/pypi/pip-date
[3]: https://badge.fury.io/py/pip-date.svg
[4]: https://badge.fury.io/py/pip-date
[5]: https://img.shields.io/badge/Maintained%3F-yes-green.svg
[6]: https://GitHub.com/E3V3A/pip-date/graphs/commit-activity
[7]: https://img.shields.io/github/last-commit/E3V3A/pip-date.svg
[8]: https://github.com/E3V3A/pip-date/commits/master "Last commits to Master branch"
[9]: http://isitmaintained.com/badge/resolution/E3V3A/pip-date.svg
[10]: http://isitmaintained.com//project/E3V3A/pip-date "Average time to resolve an issue"

A simple *Python3* CLI tool to show the installation or modification times of all your pip packages.

| STATUS: | Version | Date | Maintained? |
|:------- |:------- |:---- |:----------- |
| Working | `1.0.5` | 2022-01-23 | YES |

---

Example Output:

![Full](./docs/screen1.png)

![Full](./docs/screen2.png)


Using `pip-search <RegEx>`:

![Full](./docs/screen3.png)

---

**Q:** *What does **`pip-date`** do?*

The primary use is for finding the *time* when a certain pip package was last modified or installed. 
It is basically using one or more of: **atime, ctime** and **mtime** from the file status (*stat*) info.
This is essentially equivalent to using the \*nix *stat* command, but is handled differently on Windows. 
(See below for further details.)

Using this information, it can show you and highlight packages that may have been corrupted or outdated. 
I also has some functionality of checking packages for outdated and deprecated installation methods. 

But it can do more. Some features are:

- Highlight packages with inconsistent file modification times (*mTime*).
- Highlight package **versions** which are not conforming to the [PEP-0440](https://www.python.org/dev/peps/pep-0440/) standard.
- Highlight packages installed with an unusual package distribution *priority* given by:  **`[chk, src, bin, egg, dev]`**
- Highlight *`setuptools`* dependency packages for easy review
- Show package installation type:  with pip/wheel as **`wheel`**, and source as **`sdist`**  (**FIX!**)
- Show package installation location:   **`usr`** for `--user` and **`sys`** for *global* installations.
- Show correct file modification time, depending on OS/FS architecture (*mtime* vs *ctime*)

**Q:** *What does it **not** do?*

- Does not install packages
- Does not show dependencies
- Does not (yet) show packages in a *`virtualenv`* or *`pipenv`* envrionment (**ToDo**)
- Does not check package consistency 
- Does not show the very first time you installed a package, if it has been updated since.  
  (Althought there are left-over artifacts that may show otherwise, we don't look for these.)


**Q:** *Why is this needed?*

It probably isn't, **BUT**...

Because python packages often rely on a large number of sub-dependencies, it is very easy to accidentally 
overwrite some required dependency of one package with a different version needed by another package. 
You will never know about it, until it breaks something. One common scenario causing package corruption 
is that you have installed some package XXX using pip, but then get an OS update and install the update 
using you OS packagemanagement system, like `apt-get install XXX`, which would probably overwrite the 
*globally* installed pip package. This is especially true for beginners of python, who has not yet learned 
how to use a virtual environment, and installing evything in either the global system (default) or 
user (`--user`) environments. This may also occur when installing packages from sources, or when 
you have to run some other non-pip installers like `setup.py`, `make install` or like, 
and you don't really know what it is going to do. 


**Q:** *What else is included?*

* A script called **`pip-describe`**, that will do what *pip* doesn't, which is to show the 
full-text `long_description` for **any** PyPI package (including those not already installed).

* A script called **`pipbyday`**, that will print a simple table with:  
  `mTime/aTime` +  `package-name` + `package-version`, sorted by time.

* A script called **`pyfileinfo`**, that will show detailed file and date information 
for a given file using python's `os.stat` info.

* A script called **`pyOSinfo`**, that will print a number of *os, system* 
and *platform* variables, as seen by your Python interpreter.

* **`NEW`** A script called (hold your breath!) **`pip-search`**, that will download and 
search all of the PyPi package database for packages matching your search criteria. 
However, for colored and sexy package searches, I recommend using [pip_search](https://github.com/victorgarric/pip_search).
Also, for cool powershell wrapper of pip-search, read [this](https://github.com/E3V3A/pip-date/issues/17#issuecomment-1019325517).

**Q:** *Will I continue to support this tool?*

Sure, if it is broken, but I will not spend any more time for new features. So if you would like to add 
something just send me a PR, or at the very least, a detailed code snippet of what I need to implement.

---

### Dependencies


* [requests](https://github.com/requests/requests) - used by `pip-describe` to get PyPI info
* [lxml](https://github.com/lxml/lxml) - used by `pip-search` to parse html from PyPI

and what you already have: 
* [Python3](https://www.python.org/) 
* [pip](https://github.com/pypa/pip/)


### Installation 

There is nothing to install really. Just download the `pip-date.py` file and make sure to place it in your `PATH`.

**For pip installation:**

```bash
pip install pip-date
```


**For single file installation:**

```bash
cd /usr/bin/
wget https://github.com/E3V3A/pip-date/raw/master/pip-date
chmod 755 pip-date
```


**For developer installation:**

```bash
git clone https://github.com/E3V3A/pip-date.git
cd pip-date
pip install pip-date --user
```


### How to Run

```bash
pip-date      # When it's in your PATH
./pip-date    # When it's not in your PATH
```

---

### References:

**Time Stamps**

It's quite amusing to see how different OS's and File System's (FS) are handling file time stamps.
In the Linux world the available time stamps are called [atime](), [ctime]() and [mtime](), where 
they are generally available through the *`stat`* command. However, Windows systems doesn't have 
this commmand because NTFS is using a different way to keep track of time-stamps, so older python
versions would (and Windows OS) would report the wrong values for these. But have since been fixed.

<!-- To summarize the issue of finding the *`"last modification time"`* (*mtime*) when using Python on a 
Windows architechture, ~~we need to use *`ctime`* instead.~~ (**Has been fixed!**) Thus we use 
`platform.architecture()` to check the machine's *(bits, linkage)* tuple for the "WindowsPE" string, 
and blatantly assuming that it has a Windows FS that need *ctime*, and that anything else should 
use *mtime*. -->

Then we use: `os.path.getctime(pkg_loc)` to get the file time stamp.

For all the gory details, see: 
[here](https://linuxhandbook.com/file-timestamps/), 
[here](https://www.unixtutorial.org/atime-ctime-mtime-in-unix-filesystems/) and 
[here](https://en.wikipedia.org/wiki/MAC_times). 

---

#### Glossary:

* **`bdist`** - *"Built Distribution"*:  
	A Distribution format containing files and metadata that only need 
    to be moved to the correct location on the target system, to be 
    installed. *Wheel* is such a format, whereas distutil’s *Source 
    Distribution* is not, in that it requires a build step before it 
	can be installed. (A **"Binary Distribution"** is also a *bdist*, but 
	with additional compiled extensions.)

* **`sdist`** - *"Source Distribution"*:  
    A distribution format (usually generated using python setup.py 
    sdist) that provides metadata and the essential source files needed 
    for installing by a tool like pip, or for generating a Built 
    Distribution.

* **`egg`** - **[deprecated]**:  
    The older *Built Distribution* format introduced by *setuptools*, 
	which is being replaced by *wheel*.

* **`wheel`** - "":  
    A *Built Distribution* **format** introduced by [**`PEP-0427`**](https://www.python.org/dev/peps/pep-0427/), which is 
	intended to replace the "egg" format. A wheel (`bdist_wheel`) is a 
	ZIP-format archive with a specially formatted file name and using 
	the **`.whl`** extension. Normally, you need one wheel file for each 
	operating system and architechture. And that list can get long for big 
	projects, like *numpy*.


---

#### Recommeded or Similar Tools:

- **[pip-check](https://github.com/bartTC/pip-check/)** - Check you pip package update status with nice ANSI colored CLI
- **[pip-chill](https://github.com/rbanffy/pip-chill)** - Lists only the dependencies (or not) of installed packages
- **[pip_search](https://github.com/victorgarric/pip_search)** - Victor's amazing pip search replacement
- **[venvlink](https://github.com/np-8/venvlink)** - Using virtual environments outside of the project folder, like a boss!
---

#### Bugs and Warnings

None


#### ToDo / Help Needed

See issues marked [ToDo](https://github.com/E3V3A/pip-date/issues?q=is%3Aopen+is%3Aissue+label%3AToDo).

#### Contribution

Feel free to post issues and PR's related to this tool.  
Feel free to fork, break, fix and contribute. Enjoy!


#### Additional Badges

[![build status][11]][12] [![codecov][13]][14]
[![codecov](https://codecov.io/gh/E3V3A/pip-date/branch/master/graph/badge.svg?token=wHBDKVmtPk)](https://codecov.io/gh/E3V3A/pip-date)

---

#### License

[![GitHub license][21]][22]  
A license to :sparkling_heart:!

<sub>I use `GPLv3` because sharing code modifications is more beneficial for the world.</sub>

[11]: https://ci.appveyor.com/api/projects/status/github/pip-date/pip-date?branch=master&svg=true
[12]: https://ci.appveyor.com/project/pip-date/pip-date
[13]: https://codecov.io/gh/E3V3A/pip-date/branch/master/graph/badge.svg?token=wHBDKVmtPk
[14]: https://codecov.io/gh/pip-date/pip-date

[21]: https://img.shields.io/github/license/E3V3A/pip-date.svg
[22]: https://github.com/E3V3A/pip-date/blob/master/LICENSE.txt

CHANGES.txt
v1.0.5, 2022-01-23 -- New Release and bug fixes:modification date 

* Added `*.py` extension to all script files for Windows compatibility
  (Fixes #12 and #22)
* Fixed #21, pip-search with no arg issue
* Add sys.implementation to pyOSinfo (fixes #13)
* Bump version v1.0.5

v1.0.4, 2022-01-22 -- New Release and bug fixes:

* Fixed PR #11 for venv exception handling
* Fixed pyOSinfo uname issue.
* Fixed ANSI color capability check in is_posix()
  (May have broken on older Windows machines or Python versions.)
* Fixed wrong Windows-10 creation and modification times ([cm]Times)
  (Need feedback from native *nix users.)
* Added new script pip-search 
* Changed Shebangs from 'python3' to 'python'
* Updated README
* Bump version v1.0.4

v1.0.4, 2022-01-22 -- Sync of old work

* Synchronisation of previous work (2019-02-15) not pushed to repo.

New Release and bug fixes:
* Added __main__ function coding style
* Added __author__ etc.
* Added new search tool: pip-search
* Fixed confused [acm]times as used in Windows-10, WSL
* Some code cleanup to improve flake8 code formatting analysis:
* Removed unused imports
* Removed redundant whitespace not affecting readability
* Fixed wrong date in CHANGES.txt

v1.0.3, 2019-02-08 -- New Release and bug fixes:

* Fixed path search mechnaism for full Linux support. Fixed #6
* Fixed pyfileinfo crashing on binary files. Fixed #8
* Fixed correct time order for [acm]time
* Better terminal color-code compatibility check. Fixed #3
* Removed ANSI control code garbage for non-ANSI-compatible 
  terminals often found on Windows (CMD, PowerShell). Fixed #2
* Implemented a 'uname' check for non-uname supported OS's
* Improved IBM-437 OEM_CHARSET support for block characters. Fixed #5
* Improved usage text to pip-describe. Fixes #7
* Improved package location info: [apt,dev,sys,usr]
* Improved pyOSinfo getsitepackages() info
* Improved pipbyday to also show package location 
* Clarified table header and made it dynamic for different OS
* Added usage help text to pip-date

This release fixes all bugs that I could find and track.
But I don't have a Mac so have not been able to test it there.

Affected files:

      modified:   CHANGES.txt
      modified:   README.md
      modified:   __init__.py
      modified:   pip-date
      modified:   pip-describe
      modified:   pipbyday
      modified:   pyOSinfo
      modified:   pyfileinfo
      modified:   setup.py


v1.0.2, 2018-12-02 --  Package maintenance update with 3 new tools

    * Fix for issue #2 (windows)
    * Fixed UI PEP typo
    * Added pip-describe (requests) - get full descriptions of not installed packages
    * Added pyOSinfo - to get os,platform info
    * Added pyfileinfo - to get python-based stat info
    * General code and README cleanup
    * Updated color-highlight scheme


v1.0.1, 2018-11-26 -- Initial release


pip-date.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#----------------------------------------------------------------------
#   Author:         E:V:A
#   Date:           2018-02-15
#   Change:         2022-01-24
#   Version:        1.0.5
#   License:        GPLv3
#   URL:            https://github.com/E3V3A/pip-date/
#   Description     Show the install date of all pip-installed python3 packages
#----------------------------------------------------------------------
#  ToDo:
#   [ ] better RegEx for "pip"
#   [/] add flake8 QA ignore comments 
#   [/] fix rounding of floats in [a/c/m]Time
#   [ ] Add CLI options:
#       - [ ] '-d'          : Enable extra debug info
#       - [ ] '-e'          : Show "env" column to display 'virtualenv' name
#       - [ ] '-f'          : Force to use opposite (to detected) ctime method for FS
#       - [ ] '-n'          : Disable colors
#       - [ ] '-h'          : THIS help/usage message
#       - [ ] '-v'          : THIS program version
#       - [ ] '-t <days>'   : To highlight packages installed <days> ago
#
# NOTES:
#----------------------------------------------------------------------
#    [a/c/m]time
#    ------------------------------------------------------------------
#    On Windows (via Cygwin & Python3):
#      The creation time is:       aTime           .CreationTime === .LastAccessTime in Poweshell, but known as "access" time in Linux)
#      The modification time is:   mTime == cTime  .LastWriteTime in Poweshell
#
#    On Linux:
#      The creation time is:       cTime
#      The modification time is:   mTime
#      The access time is:         aTime           (normally not used)
#
#    ==> For seeing last modification time, use "cTime" on Windows FS's, and "mTime" on *linux FS's
#----------------------------------------------------------------------
#  References:
#   [1] https://linuxhandbook.com/file-timestamps/
#   [2] https://www.unixtutorial.org/atime-ctime-mtime-in-unix-filesystems/
#----------------------------------------------------------------------
import re, os, sys, platform    # noqa: E401
#import subprocess
import site, pkg_resources      # noqa: E401
from datetime import datetime
from datetime import timedelta
#from time import strftime

__author__ = "E:V:A (E3V3A)"
__copyright__ = "GPLv3 2022"
#__credits__ = ["https://github.com/E3V3A/pip-date/"]
__version__ = '1.0.5'

#----------------------------------------------------------
# OS Check-1
#----------------------------------------------------------
# Apparently for:
# python.exe -c "import os,sys; print('TERM=%s' % os.getenv('TERM'));"
# python -c "import os; print('\n'.join([os.name, os.sys.platform]));"
#   PowerShell/CMD Windows python:  TERM=None
#   PowerShell/CMD Cygwin python:   TERM=cygwin
#   WSL                             TERM=xterm-256color
#----------------------------------------------------------
def is_posix():
    noco = ('dumb', 'xtermm', 'xterm-mono')         # dumb = No VT,  'others' --> $PSStyle.OutputRendering = PlainText
    px_term = os.getenv('TERM')                     # [cygwin, xterm, xterm-color, xterm-256color]
    px_name = os.name                               # [posix, nt, ...]
    px_plat = sys.platform                          # [linux, cygwin, win32]

    if px_term in noco: 
        return False
    if ((px_term == 'None') or (px_term == '')):    # For native Windows "consoles" often return "None", since ="".
        if ( (px_name == 'nt') and (px_plat == 'win32') ):
            return True
        else: 
            return False
    # Assume we have a color term
    return True

#----------------------------------------------------------
# OS Check-2
#----------------------------------------------------------
# We need to test how [a/c/m]time works on the OS
def isWinFS():
    if platform.architecture()[1] == "WindowsPE":
        print("Using cTime for WindowsPE\n")
        return True
    else:
        print("Using mTime for Linux FS\n")
        return False

#----------------------------------------------------------
# Text Coloring
#----------------------------------------------------------
# Usage:  print(yellow("This is yellow"))
def color(text, color_code):
    #if self.nposix:
    if not is_posix():
        return text
    # for brighter colors, use "1;" in front of "color_code"
    bright = ''  # '1;'
    return '\x1b[%s%sm%s\x1b[0m' % (bright, color_code, text)

def red(text):    return color(text, 31)            #                   # noqa
def green(text):  return color(text, 32)            # '1;49;32'         # noqa
def bgreen(text): return color(text, '1;49;32')     # bright green      # noqa
def orange(text): return color(text, '0;49;91')     # 31 - looks bad!   # noqa
def yellow(text): return color(text, 33)            #                   # noqa
def blue(text):   return color(text, '1;49;34')     # bright blue       # noqa
def purple(text): return color(text, 35)            # aka. magenta      # noqa
def cyan(text):   return color(text, '0;49;96')     # 36                # noqa
def white(text):  return color(text, '0;49;97')     # bright white      # noqa

#----------------------------------------------------------
# Print Usage
#----------------------------------------------------------
def usage():
    print(" Usage:  %s\n" % os.path.basename(__file__))
    print(" This will return a detailed sorted list of all your installed packages.")
    print(" The command doesn't take any arguments, and is part of the pip-date")
    print(" package. Other commands includeded in this package are:  pipbyday,")
    print(" pip-describe, pyfileinfo and pyOSinfo.\n")
    print(" Please file any bug reports at:")
    print(" https://github.com/E3V3A/pip-date/\n")
    print(" Version:  %s" % __version__)
    print(" License:  GPLv3\n")
    sys.exit(2)

#----------------------------------------------------------
# Print Warning
#----------------------------------------------------------
def print_warning():
    print('\n')
    print('-'*60)
    print(' WARNING!')
    print(' You are missing out on important color coded information!')
    print(' This is because you are probably using a Windows console')
    print(' that is not fully supporting ANSI color sequencies.')
    print(' For best experience, either run this in Cygwin or WSL,')
    print(' or install WinPty, ConEmu or a PowerShell version >6.1.')
    print(' If you do have a POSIX compatible color terminal, then')
    print(' make sure your TERM environment variable is set.')
    #print(' (Usally to \"xterm\".)')
    print('-'*60)

#----------------------------------------------------------
# Print Color Legend
#----------------------------------------------------------
# See:
#   https://github.com/PowerShell/PowerShell/issues/8409
#   https://en.wikipedia.org/wiki/Code_page_437
#   https://en.wikipedia.org/wiki/Box-drawing_character
#   https://en.wikipedia.org/wiki/Block_Elements
# Let's try:
#   2585,               # Look best but is not part of cp437 and thus font dependent & not widely available
#   2580, 25A0, 2588    # IBM-437
#----------------------------------------------------------
def print_legend():
    #cc = u'\u2585' # Unicode Character for a "5/8th box"           # (U+2585) is not part of IBM-437
    #cc = u'\u2588' # Unicode Character for a "full box"            # (U+2588) is part of IBM-437
    #cc = u'\u25A0' # Unicode Character for a "black square"        # (U+25A0) is part of IBM-437
    cc = u'\u2580'  # Unicode Character for a "Upper half block"    # (U+2580) is part of IBM-437
    print("  {} = ERROR (preventing package processing)".format(red(cc)))
    print("  {} = Using a Bad, Deprecated or Non-Standard installation Path".format(purple(cc)))
    print("  {} = Possibly Multiple installations (differing file times)".format(yellow(cc)))
    print("  {} = Recently Changed / Installed (in last 7 days)".format(cyan(cc)))
    print("  {} = Non-PEM-compliant Version string (PEP-0440) | ~/.local install".format(green(cc)))
    print("  {} = A 'setuptools' dependency package".format(blue(cc)))

#----------------------------------------------------------
# Helper Functions
#----------------------------------------------------------
def safe_name(name):
    # Replace runs of non-alphanumeric characters with a single '-'.
    return re.sub('[^A-Za-z0-9]+', '-', name)

def safe_version(version):
    # Convert an arbitrary string to a standard version string
    version = version.replace(' ', '.')
    return re.sub('[^A-Za-z0-9.]+', '-', version)

def to_filename(name):
    # Replace any '-' characters with '_'.
    return name.replace('-', '_')

def test_loc(loc):
    # Test package location to give us some idea of what type of install it came with.
    #  .local           : are usually local user installs in:  $HOME/.local/lib/pythonX.Y/site-packages/...
    #  site-packages    : are usually system user installs (sudo)
    #  dist-packages    : are usually system package-manager installs (apt)
    #  /PATH/           : are usually developer installs using "pip install ."
    if '.local' in loc:
        ploc = bgreen('usr')        # user (unprivileged local install)
    elif 'dist-packages' in loc:
        ploc = 'apt'                # system (apt package-manger installed)
    elif 'site-packages' in loc:
        ploc = 'sys'                # system (user sudo installed)
    else: 
        #ploc = red(loc)            # show actual path
        ploc = red('dev')           # dev (user development install via "pip install .")
    return ploc

def pre2txt(pre):
    # Distribution "precedence" constants:  (../pkg_resources/__init__.py)
    # EGG_DIST, BINARY_DIST, SOURCE_DIST, CHECKOUT_DIST, DEVELOP_DIST  :  [3,2,1,0,-1]
    # { 'egg': 3, 'bin': 2, 'src': 1, 'chk': 0, 'dev': -1 }
    # However, this seem poorly implemented since most packages show "-1" or 3.
    d = ['chk', 'src', 'bin', 'egg', 'dev']
    return d[pre]

def is_canonical(version):
    # Check PEP-0440 Version string compliance:
    # https://www.python.org/dev/peps/pep-0440/
    canrex = r'^([1-9]\d*!)?(0|[1-9]\d*)(\.(0|[1-9]\d*))*((a|b|rc)(0|[1-9]\d*))?(\.post(0|[1-9]\d*))?(\.dev(0|[1-9]\d*))?$'
    return re.match(canrex, version) is not None

def pkgcol(pkgarr):
    #----------------------------------------------------------
    # The <package_name> color require special treatment, because
    # of sorting on key position and getting ljust space.
    # https://packaging.python.org/key_projects/
    cygset  = ['setuptools', 'appdirs', 'packaging', 'pyparsing', 'six']    # Cygwin python3-setuptools dependency packages
    #cygset += ['wheel', 'virtualenv', 'pipenv', 'pip']                      # ...some additional essentials
    #cygset += ['scikit-build', 'distlib']                                   # ...some additional essentials
    # NOTE!
    #   We can't use "pip" because the we're only checking if the string is present in line,
    #   Thus anything with "pip" in it would be caught, so we need a smarter RE here.
    #----------------------------------------------------------
    line = ''
    for i in range(len(pkgarr)):
        line = pkgarr[i]
        for pname in cygset:
            # ToDo: only replace if in 1st word
            if pname in line:
                #pkgarr[i] = re.sub(r'^[a-zA-Z0-9_\-]+', blue(pname), line, 1)
                rx = line.find(pname)
                if rx < 20 and rx != '-1':
                    pkgarr[i] = line.replace(pname, blue(pname), 1)
                    break
    return pkgarr

#----------------------------------------------------------
# MAIN
#----------------------------------------------------------
def main_func():

    print()
    debug    = 0
    #nposix   = 0
    pcnt     = 0
    pkg      = []

    #---------------------------------------
    # CLI arguments
    #---------------------------------------
    narg = len(sys.argv) - 1
    if narg >= 1:
        #pkg = sys.argv[1]
        usage()
    #if is_posix():
    #    nposix = 1
    #---------------------------------------

    # Check [a/c,m]time availability
    useWinStat = False      # Linux ELF based FS system
    if isWinFS():
        useWinStat = True  # WindowsPE FS system

    #---------------------------------------
    #  MacOS: ??
    #  Cygwin: native python, we only have 1 location:
    #       /usr/lib/python3.6/site-packages
    #  Linux Mint (19.1): we have several (3)
    #       /usr/local/lib/python3.6/dist-packages
    #       /usr/lib/python3/dist-packages
    #       /usr/lib/python3.6/dist-packages
    # Also add the unprivileged user's local package location:
    #       $HOME/.local/lib/python3.6/site-packages/
    #---------------------------------------
    try:
        site_loc = site.getsitepackages()                   # [...]
        if debug: print("site_locs (site): ", site_loc)
        site_loc += [site.getusersitepackages()]            # add $HOME/.local/lib/python3.6/site-packages/
        if debug: print("site_locs (all): ", site_loc)
    except AttributeError:
        site_loc = site.USER_SITE

    for d in pkg_resources.working_set:

        try:
            pkg_name = d.project_name                                                               #
            pkg_ver  = d.version                                                                    #
            #pkg_loc  = d.location                                                                  # NOT always a file!
            pkg_typ  = 'n/a' # "wheel" if d.location.is_wheel else "sdist"                          # 'Type'

            if debug:
                pkg_pre  = d.precedence                                                             # 'Prec' [-1..3]
            else:
                pkg_pre  = pre2txt(d.precedence) if (d.precedence != -1) else ''                    # 'Prec' string

            pkg_ins  = d.get_metadata('INSTALLER').strip() if d.has_metadata('INSTALLER') else ''   # get_metadata_lines() '???'
            pkg_whl  = d.get_metadata('WHEEL').strip() if d.has_metadata('WHEEL') else ''           # get_metadata_lines() '???'
            # ^^^^^ This often have multiple lines, we need to format:
            if pkg_whl:
                pw = pkg_whl.split('\n')
                pw = '\n    ' + '\n    '.join(pw,)
                pkg_whl = pw

        except ValueError as e:
            print(red("ERROR:") + " %s" % e)

        #---------------------------------------
        # Get the correct package location
        #---------------------------------------
        # Because d.location doesn't return a file, but only a directory,
        # for certain packages, we also check the "module directory" ???
        try:
            mod_dir = next(d._get_metadata('top_level.txt'))    # module_dir
            pkg_loc = os.path.join(d.location, mod_dir)         #
            os.stat(pkg_loc)                                    #

        except (StopIteration, OSError):
            try:
                pkg_loc = os.path.join(d.location, d.key)
                os.stat(pkg_loc)
            except OSError:
                pkg_loc = d.location
        #---------------------------------------

        if debug:
            print('-'*40)
            print("pkg_loc:  %s: %s  (%s)" % (pkg_name.ljust(20,' '), pkg_loc, pkg_pre))    # pkg_pre always empty ??
            print("pkg_typ:  %s" % pkg_ins)
            print("pkg_whl:  %s" % pkg_whl)
            print("pkg_ins:  %s" % pkg_ins)

        # A work-around for packages with deprecated location(s):
        if ".egg" in pkg_loc: 
            print(purple("Found Bad Path Location for:") + "  %s" % white(pkg_name))
            print("Package Location found at:    %s" % (pkg_loc))

        #---------------------------------------
        # Getting OS Dependent TimeStamps
        #---------------------------------------
        # NOTE:
        # (1) [acm]time as used in variable names HERE, is true for LinuxFS,
        #     but swapped for WindowsFS's
        # (2) In a WindowsPE based FS:
        #       (a) The true "creation" time is the (python stat) "atime"
        #       (b) The true "modification" time is the (python stat) "mtime"=="ctime"
        # (3) In a ELF based LinuxFS:
        #       aTime = access time ...     - Rarely used because of FS performance
        #       mtime = modification time   -
        #       ctime = creation time       - is the "real" last modification time on windows
        #---------------------------------------
        if debug: print("pkg_loc:  %s" % pkg_loc)

        if os.path.exists(pkg_loc):
            if useWinStat:
                tsc = os.path.getctime(pkg_loc)     # WindowsFS:    ctime:  "creation time"
                tsm = os.path.getmtime(pkg_loc)     # WindowsFS:    mtime:  "last modified"
                #tsm = os.path.getatime(pkg_loc)    # WindowsFS:    atime:  "last accessed"
            else:
                tsc = os.path.getctime(pkg_loc)     # LinuxFS (ctime)
                tsm = os.path.getmtime(pkg_loc)     # LinuxFS (mtime)
        else:
            #print(red("Skipping Bad Path of:") + " %s: \t%s" % (pkg_name, pkg_path))   # to_filename(pkg_name))
            print(red("Skipping Bad Path of:") + " %s: \t%s" % (pkg_name, pkg_loc))     # to_filename(pkg_name))
            continue

        #------------------------------------------------------------------------------------
        # Processing Time Stamps
        #------------------------------------------------------------------------------------
        # Logic:  
        #   1. IF  ( mTime > (K * cTime) )   THEN  highlight mTime  ELSE  don't show  # (1)
        #   2. IF  ( cTime < '1-week-ago' )  THEN  highlight cTime                    # (2) 
        #
        #------------------------------------------------------------------------------------
        pkg_ctime = datetime.fromtimestamp(tsc).strftime("%Y-%m-%d  %H:%M:%S").strip()  # str
        pkg_mtime = datetime.fromtimestamp(tsm).strftime("%Y-%m-%d  %H:%M:%S").strip()  # str

        # NOTE:  If TS differs by < 1 second, it will not show. To do so, test:  tsc == tsm
        max_tdelta = 60                         # Max allowed time difference: ~60s
        tdelta = abs(tsc - tsm)                 # Calculate time difference
        #if pkg_ctime != pkg_mtime:             # This is too restrictive (OS need seconds to install)
        if tdelta > max_tdelta:                 #
            pkg_mtime = yellow(pkg_mtime)       # (1) Highlight packages with different creation vs modification Times
        else:
            pkg_mtime = ''                      # Skip those where: mTime ~= cTime

        ctNow = datetime.now()                  # Time: "now"
        ct7dy = ctNow - timedelta(days=7)       # Time: 1-week-ago
        pctim = datetime.fromtimestamp(tsc)     # Time: file cTime (Time Stamp)
        if pctim > ct7dy:
            pkg_ctime = cyan(pkg_ctime)         # (2) Highlight packages recently created (cTime < 1-week-ago)
        #------------------------------------------------------------------------------------

        if pkg_pre:
            pkg_pre = yellow(pkg_pre) + " "     # [egg,...]

        if not is_canonical(pkg_ver):
            pkg_ver = green(pkg_ver) + " "*3    # green + ugly color-code-length hack when using colors...

        # ugly test here... must be a better way
        if 'bdist_wheel' in pkg_whl:
            pkg_typ = 'wheel'
        else:
            #pkg_typ = white('sdist')
            pkg_typ = orange('sdist')

        pkg_loc = test_loc(pkg_loc)

        pcnt += 1
        pkg  += ["{:20}   {:<20}   {:<20}   {:<16}   {:6}   {:<4}   {:5}   {:3}".format(pkg_name.ljust(20,' '), pkg_ctime, pkg_mtime, pkg_ver, pkg_ins, pkg_pre, pkg_typ, pkg_loc)] # noqa

    if useWinStat:
    #   header_str = "{:20}   {:20}   {:20}   {:16}   {:6}   {:4}   {:5}   {:3}".format('Package'.ljust(20, ' '), 'LastModified (mTime)', 'FirstSeen (aTime)', 'Version', 'Inst', 'Prec', 'Type ', 'Loc')
        header_str = "{:20}   {:20}   {:20}   {:16}   {:6}   {:4}   {:5}   {:3}".format('Package'.ljust(20, ' '), 'Installed (cTime)', 'LastModified (mTime)', 'Version', 'Inst', 'Prec', 'Type ', 'Loc')
    else:
        header_str = "{:20}   {:20}   {:20}   {:16}   {:6}   {:4}   {:5}   {:3}".format('Package'.ljust(20, ' '), 'LastModified (mTime)', 'FirstSeen (cTime)', 'Version', 'Inst', 'Prec', 'Type ', 'Loc')

    hlen = len(header_str)
    print('\n' + header_str)

    print("-"*hlen)
    spkg = sorted(pkg, key=str.lower)
    spkg = pkgcol(spkg)
    print('\n'.join(spkg))

    print("-"*hlen)
    print_legend()

    print("-"*hlen)
    print("Found %d packages." % pcnt)

    #if not nposix:
    if not is_posix():
        print_warning()

    print("\nDone!")


if __name__ == "__main__":
    main_func()
    sys.exit(0)
